src/Data/Linear/Vector.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+-- |
+--   Module: Vector
+--   Description: A fast dynamic, mutable vector.
+--
+--   - TODO:
+--
+--   - Add doc.
+--   - XXX: is doubleVec functional???
+--
+--   - Functor Instance
+--   - Dupable instance
+module Data.Linear.Vector
+  ( Vector,
+    singleton,
+    read,
+    write,
+    snoc,
+    length,
+  )
+where
+
+import Data.Unrestricted.Linear (Unrestricted (..))
+import GHC.Exts
+import qualified Unsafe.Linear as Linear
+import Prelude hiding (length, read)
+import Data.Linear.Internal.MutableArray
+
+-- # Core data types
+-------------------------------------------------------
+
+type MutArr a = MutableArray# RealWorld a
+
+-- | A dynamic mutable vector.
+data Vector a where
+  -- | Vec (memory allocated, current length) marray
+  Vec :: (Int, Int) -> MutArr a -> Vector a
+
+-- # API: Construction, Mutation, Queries
+-------------------------------------------------------
+
+-- | Singleton vector
+singleton :: a -> Vector a
+singleton elem = Vec (1,defaultSize) (unsafeNewMutArr defaultSize elem)
+
+-- | Length of the vector
+length :: Vector a #-> (Int, Vector a)
+length = Linear.toLinear unsafeLength
+  where
+    unsafeLength :: Vector a -> (Int, Vector a)
+    unsafeLength v@(Vec (len, _) _) = (len, v)
+
+-- | Insert at the end of the vector
+snoc :: Vector a #-> a -> Vector a
+snoc (Vec (len, size) ma) x
+  | len < size = write (Vec (len + 1, size) ma) len x
+  | otherwise = write (doubleVec (Vec (len + 1, size) ma)) len x
+
+-- | Write to an element already written to before.
+-- Note: this will not write to elements beyond the current
+-- length of the array.
+write :: Vector a #-> Int -> a -> Vector a
+write = Linear.toLinear writeUnsafe
+  where
+    writeUnsafe :: Vector a -> Int -> a -> Vector a
+    writeUnsafe v@(Vec (len, size) mutArr) ix val
+      | indexInRange len ix = case writeMutArr mutArr ix val of () -> v
+      | otherwise = error "Index not in range"
```

Comment:
I see that you haven't put the `HasCallStack` constraints yet. Let's do that next.